Add a new editor theme

YeldhamDev opened the pull request:
<img width="1920" height="1128" alt="Screenshot_20251023_132202" src="https://github.com/user-attachments/assets/3f95136f-58ee-4b66-a117-415cf83719df" />
<br><br>

Is with great pleasure that I present you, Godot's soon-to-be new default theme!

This is a 1:1 port of @passivestar's [Minimal Theme](https://github.com/passivestar/godot-minimal-theme?tab=readme-ov-file), with himself also helping us behind the curtains in order to facilitate the porting work. The editor font used is now [Inter](https://rsms.me/inter/), and the new theme also changes the editor's base color and accent. While the blue tint had become a staple of Godot’s look over the past years, it caused issues such as distorting the perceived white balance of the viewport. Switching to a neutral gray resolves this.

But for those who still liked the older look better, fret not, As **the old theme and color preset is still available**, for you to pick! They can be accessed by choosing "Classic" in the `interface/theme/style` and `interface/theme/preset` settings.

While this PR is in a merge-able state, some small things still need to be worked on:
- <s>Nested `TabContainer`s blend with each other.</s>
- <s>The disabled tab style is a placeholder, while waiting for passive to implement it.</s>
- `PopupMenu`s still look very squarish, as #106324 still needs to be resolved.

Feedback is more than welcome, but please, keep it civil. Topics involving theming and design are very subjective, so the more objective information you can provide about your issues the easier it will be to address them in future PRs.

YeldhamDev committed:
Add a new editor theme

Co-authored-by: passivestar <60579014+passivestar@users.noreply.github.com>

KoBeWi commented:
As I already mentioned in our internal discussion, the theme makes various elements lose distinction by blending their background with surroundings. Some examples:

Scene dock (no background, the tree looks like it was disabled or something)
|Old|New|
|-|-|
|<img width="278" height="456" alt="image" src="https://github.com/user-attachments/assets/ba869ed7-fbdd-442c-b37c-3547b659344c" />|<img width="277" height="476" alt="image" src="https://github.com/user-attachments/assets/6464a3c3-22a6-4cc3-a8ea-03562ef60387" />|

Debugger (there is a big space that looks completely empty)
|Old|New|
|-|-|
|<img width="1371" height="296" alt="image" src="https://github.com/user-attachments/assets/a8d987fa-8534-4d8f-9976-a3592d09c250" />|<img width="1288" height="304" alt="image" src="https://github.com/user-attachments/assets/84190727-d7ed-4372-90ad-d337595d37b5" />|

Inspector (categories blend with properties)
|Old|New|
|-|-|
|<img width="288" height="415" alt="image" src="https://github.com/user-attachments/assets/6596d252-d136-4794-918a-8f23842ce438" />|<img width="386" height="432" alt="image" src="https://github.com/user-attachments/assets/94d71607-3451-415a-9552-17cd344b8933" />|

Create dialog (the list just starts out of nowhere)
|Old|New|
|-|-|
|<img width="770" height="506" alt="image" src="https://github.com/user-attachments/assets/3d5b4548-1257-4482-932b-86d9491f395c" />|<img width="753" height="546" alt="image" src="https://github.com/user-attachments/assets/8cf4b3cb-39c3-434e-aeae-c2a68ab8ba0f" />|

This makes the editor less readable. It needs to be resolved before the theme can be considered for default.

YeldhamDev commented:
> how does this interact with light theme? is there a light variant of this or is the light theme going to revert to the default spacings?

The current color presets are all there, including the light theme.

bruvzg commented:
> What about visual consistency across languages compared to Noto Sans?

Metrics seems similar, so should not be an issue. But we probably should allow overriding language specific fonts (not as part of this PR, I'll look at it a bit later).

YeldhamDev commented:
@KoBeWi That actually seems to be a regression from another PR. Notice how the theme names don't have spaces anymore in your video, while the code that checks them does.

DeeJayLSP commented:
Related: https://github.com/godotengine/godot-proposals/discussions/9012

F3der1co commented:
These are difficult to discuss as it can quickly get opinionated and subjective, but I feel like the value of the grey color is a bit too dark. 
This causes a similar issue to the white balance but for value contrast. It also makes switching to different content, like previewing the game, to have potential for a larger sudden increase in brightness, which can be painful/take time to adjust. 
So I would suggest moving the default grey from the current 15% to 20%.

DeeJayLSP commented:
> Related: [godotengine/godot-proposals#9012](https://github.com/godotengine/godot-proposals/discussions/9012)

If it isn't much to suggest, would be good to have `ss02` (disambiguation, enables uppercase i with serif, lowercase L with tail and slashed zero) defined as it fixes the biggest complain every FOSS project has when it comes to Inter.

DeeJayLSP commented:
I believe this PR changes too much things at once.

I did some work replacing the default font before, so I decided to open #111140 to track it separately.

YeldhamDev commented on editor/themes/editor_fonts.cpp:
```diff
@@ -188,8 +188,8 @@ void editor_register_fonts(const Ref<Theme> &p_theme) {
 
-	Ref<FontFile> default_font_bold = load_internal_font(_font_NotoSans_Bold, _font_NotoSans_Bold_size, font_hinting, font_antialiasing, true, font_subpixel_positioning, font_disable_embedded_bitmaps, false);
-	Ref<FontFile> default_font_bold_msdf = load_internal_font(_font_NotoSans_Bold, _font_NotoSans_Bold_size, font_hinting, font_antialiasing, true, font_subpixel_positioning, font_disable_embedded_bitmaps, font_allow_msdf);
+	Ref<FontFile> default_font_bold = load_internal_font(_font_Inter_Bold, _font_Inter_Bold_size, font_hinting, font_antialiasing, true, font_subpixel_positioning, font_disable_embedded_bitmaps, false);
```
Adding context from https://github.com/godotengine/godot/pull/111140#issuecomment-3358606096: Variable is heavier than both Regular and Bold combined.

HexagonNico commented:
I would like to add that the inspector with this theme doesn't have much feedback as to what is being clicked, especially in the categories.

https://github.com/user-attachments/assets/f6b43f75-fc9b-4eac-8c1b-0bfba547db2b

Another issue I have is that since the nodes in the scene view are closer together, the icons on the right are also a lot closer and I guess they may be harder to click on certain monitors.

| Vertical spacing is bigger | Horizontal spacing is bigger |
|--------------------------------------|------------------------------------------|
| <img width="367" height="312" alt="old_theme" src="https://github.com/user-attachments/assets/6afeafac-f8da-48e0-8852-1aa4ed62dad1" /> | <img width="369" height="312" alt="new_theme" src="https://github.com/user-attachments/assets/4446ccc7-c575-428f-8b48-e543fabcb980" />

YeldhamDev commented:
@Giganzo Try the 4.6 branch of the theme.

ettiSurreal commented:
> > Related: [godotengine/godot-proposals#9012](https://github.com/godotengine/godot-proposals/discussions/9012)
> 
> If it isn't much to suggest, would be good to have `ss02` (disambiguation, enables uppercase i with serif, lowercase L with tail and slashed zero) defined as it fixes the biggest complain [certain FOSS projects have]

I'd also suggest using the OpenType features Blender uses (which I listed on my original discussion on switching to Inter, as I think they really improve on readability.
```
cv02 - Open four
cv03 - Open six
cv04 - Open nine
cv05 - Lowercase L with tail
cv08 - Uppercase I with serif
cv09 - flat top three
tnum - Evenly spaced numbers, makes the interface less "jumpy". (example gif on the original discussion)
```

MarioBossReal commented:
> As I already mentioned in our internal discussion, the theme makes various elements lose distinction by blending their background with surroundings. Some examples:

I know that theming is subjective, but to me all of the "new" images you supply are a vast improvement over the default theme. The default theme is far too busy, pointing out categories blending with properties in the inspector (among other things) does not resonate with me as a bad thing, it's in fact perfectly readable without the eyesore of having multiple slightly different shades of gray as text backgrounds when the bold white text and dropdown arrow prefix are arguably just the right amount of visual distinction when contrasted with the dimmer regular font of the properties. 

Mind you I'm not saying the minimal theme is perfect or anything, but the main features it provides are both better use of available screen space, and much less visual clutter which to me improves readability greatly.

Giganzo commented:
> @Giganzo Try the 4.6 branch of the theme.

Keyboard focus is different for Line/TextEdit between this PR and 4.6 branch when I tested now. This PR has no focus ring and the color contrast change is very low, so hard to see.

Giganzo commented:
When button in the tree are hovered you can see that they are higher then the row itself and the icon looks like it is aligned to top of the button (looks like it only is visible when Display Scale is 100%):
<img width="365" height="466" alt="Screenshot_20251003_123855" src="https://github.com/user-attachments/assets/29d3f70c-4013-4ce0-8d17-71402f836290" />

`p_theme->set_constant("v_separation", "Tree", Math::pow(p_config.base_margin * 0.2 * EDSCALE, 3));`

one option is to change it to use the same margin as p_config.flat_button_pressed:

`p_theme->set_constant("v_separation", "Tree", p_config.base_margin * 0.9 * EDSCALE);`

<img width="365" height="466" alt="Screenshot_20251003_123750" src="https://github.com/user-attachments/assets/8a7f68e7-3858-4d2d-9fb7-3652b45ba106" />

F3der1co commented:
> those light gray panels provide very low contrast to the text.. the text is hard to read which is mostly of the same color i don't know if it can be changed with presets, i'm not used to this theme but as is those panels are too light for a dark theme. colors aside i like most of the design choices. thanks for the hard work!

I quickly checked the value kontrast, here given in percent (0 = black 100 = white, HSV space)
The new theme seems to have better value contrast:
<img width="367" height="312" alt="image" src="https://github.com/user-attachments/assets/464b6c48-cec2-449f-b586-f8e26a394c10" />
<img width="369" height="312" alt="image" src="https://github.com/user-attachments/assets/eea05918-08a2-4b95-9d3f-037addee089c" />

arkology commented:
> those light gray panels provide very low contrast to the text.. the text is hard to read which is mostly of a similar color

I'm 100% agree with you.
Just look at filters.

https://github.com/user-attachments/assets/5f29daff-1112-46eb-821a-98d57f10dd18

Also looks like something is wrong with margins because new theme is wider. In above video  right part of the editor goes out of window.
Margins in tabs take too much space.

Giganzo commented:
> The new theme seems to have better value contrast:
> (keep in mind the screenshot some people posted here didn't have the correct colors)

@F3der1co don't think that image have the correct colors.

LineEdit has very low contrast with background, 1.1:1. It fails WCAG 2.1 requires a contrast ratio of at least 3:1 for graphics and user interface components

YeldhamDev commented:
Some updates:
- Fixed bugs related to default colors and corner values.
- Adjusted tabs to take less space.
- Added some contrasting elements to some trees and nested tabs.

domdfcoding commented:
Further to what @popcar2 said I think VS Code and JetBrains IDEs all give a choice of (for them several) themes in the initial setup, so it is reasonably common to do.

DeeJayLSP commented:
> For context: most grays that can be seen in UIs are actually not true neutral grays. Try using a color picker on dark GitHub's background; you'll see it's slightly tinted blue. These tinted grays tend to look more pleasing[1](#user-content-fn-1-69bdd0b0fe582a21b3f5f92fb8c20621), but they can skew the white balance a bit. Blender has true neutral grays in comparison (R/G/B are always identical).

Just a reference with popular themes (how many times Green and Blue are higher than Red):
| Theme        | G vs R | B vs R |
|--------------|--------|--------|
| One Dark Pro | 1.1    | 1.3    |
| Dracula      | 1.05   | 1.35   |
| Nord         | 1.13   | 1.39   |
| GitHub       | 1.14   | 1.38   |

groud commented on editor/settings/editor_settings.cpp:
```diff
@@ -580,18 +580,19 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	EDITOR_SETTING(Variant::BOOL, PROPERTY_HINT_NONE, "interface/theme/draw_extra_borders", false, "")
-	EDITOR_SETTING(Variant::FLOAT, PROPERTY_HINT_RANGE, "interface/theme/icon_saturation", 1.0, "0,2,0.01")
+	EDITOR_SETTING(Variant::FLOAT, PROPERTY_HINT_RANGE, "interface/theme/icon_saturation", 2.0, "0,2,0.01")
 	EDITOR_SETTING(Variant::FLOAT, PROPERTY_HINT_RANGE, "interface/theme/relationship_line_opacity", 0.1, "0.00,1,0.01")
```
I am checking the settings here, it seems there's no way to force-show the relationship lines to be visible. Even if the default is to hidden, there should be a way to have them visible.
When they got implemented, we could see the community was pretty divided on that option, so it needs to stay a settings.

YeldhamDev replied:
Done.

DeeJayLSP commented on editor/settings/editor_settings.cpp:
```diff
@@ -580,18 +580,20 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
+	EDITOR_SETTING_BASIC(Variant::COLOR, PROPERTY_HINT_NONE, "interface/theme/base_color", Color(0.14, 0.14, 0.14), "")
+	EDITOR_SETTING_BASIC(Variant::COLOR, PROPERTY_HINT_NONE, "interface/theme/accent_color", Color(0.34, 0.62, 1.0), "")
 	EDITOR_SETTING_BASIC(Variant::BOOL, PROPERTY_HINT_NONE, "interface/theme/use_system_accent_color", false, "")
 	EDITOR_SETTING_BASIC(Variant::FLOAT, PROPERTY_HINT_RANGE, "interface/theme/contrast", 0.3, "-1,1,0.01")
```
Fun fact: If you compare the contrast using the base colors of other engines, you will find a value of 0.36 on GameMaker, and 0.37 on both Unity and Unreal.

Just a suggestion, though, but a slightly higher contrast than the default could further improve usability.

DeeJayLSP commented on editor/settings/editor_settings.cpp:
```diff
@@ -580,18 +580,20 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "interface/theme/icon_and_font_color", 0, "Auto,Dark,Light")
-	EDITOR_SETTING_BASIC(Variant::COLOR, PROPERTY_HINT_NONE, "interface/theme/base_color", Color(0.2, 0.23, 0.31), "")
-	EDITOR_SETTING_BASIC(Variant::COLOR, PROPERTY_HINT_NONE, "interface/theme/accent_color", Color(0.41, 0.61, 0.91), "")
+	EDITOR_SETTING_BASIC(Variant::COLOR, PROPERTY_HINT_NONE, "interface/theme/base_color", Color(0.14, 0.14, 0.14), "")
```
I know this is meant to be almost a 1:1 port of Minimal Theme (this color corresponds to `#242424`, but I'd like to compare with other engines and popular color themes, so:

- Unreal Engine: `#2f2f2f`
- GameMaker: `#323232`
- Unity: `#383838`
- CryEngine: `#393939`
- Cocos2d-x: `#474747`

I made a previous comment about popular text editor color themes and would like to use them as an example too. But neutral colors are important, so I got their RGB means instead:

- Solarized Dark (base03): `#292929`
- Xcode Dark: `#2c2c2c`
- One Dark Pro/Dracula/GitHub/Tomorrow: `#2d2d2d`
- Solarized Dark (base02): `#2f2f2f`
- Nord: `#363636`

Personally, I would bump this up at least to `#292929`.

arkology replied:
> * Unreal Engine: `#2f2f2f`
> * GameMaker: `#323232`
> * Unity: `#383838`
> * CryEngine: `#393939`
> * Cocos2d-x: `#474747`

For me they look like black, black, black, black, and black🤣

YeldhamDev commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -40,13 +40,10 @@
+#include "editor/themes/theme_classic.h"
+#include "editor/themes/theme_default.h"
 #include "scene/resources/texture.h"
+#include "scene/scene_string_names.h"
```
This one is still needed.

AThousandShips replied:
Indeed didn't notice `scene/gui/graph_edit.h` was removed

KoBeWi commented:
> I thought about that, but is such an option really needed to be present in the quick settings? Sounds like something you would touch once and then never again.

Then at least rename the Interface Theme to something better.

YeldhamDev commented:
Changes made.

Calinou commented:
I gave this PR a test run again, it's in a pretty good state now.

In terms of design/UX, these are my only remaining concerns before merging:

- The height of each Tree item (21 pixels per item at 100% editor scale) still feels too low by default. It should be increased to match PopupMenu's item height at least (24 pixels per item):

<img width="547" height="319" alt="image" src="https://github.com/user-attachments/assets/d9e041d6-3c1f-475f-addb-12bce1165128" />

This is also noticeable in the Project/Editor Settings, where the list of sections appears much more busy than the settings themselves:

<img width="384" height="295" alt="image" src="https://github.com/user-attachments/assets/f3019c7e-890f-48be-a1f1-edea82f8f261" />

The current height is well-suited for the Compact preset, but even then, it might need one more pixel to be vertically centered.

- The inspector lacks visible hover feedback for foldable sections.
- The background for selected inspector properties is barely visible:

<img width="292" height="204" alt="image" src="https://github.com/user-attachments/assets/efe0e28b-e24e-4490-9c38-9d286a4c35fb" />

I regularly rely on this background color to visually "bookmark" the property I last edited, so it makes usability worse for me.

The rest can be addressed in follow-ups.

DeeJayLSP commented:
> The problem is that we are still in Godot 4, so that would also be confusing. And "Godot Blue" sounds weird because it would be the only preset with "Godot" before the color name in there. Maybe just "Blue"?

Note that [Godot 3 uses these exact same colors](https://github.com/godotengine/godot/blob/f85956e45e775f312a5ffdbe6c060f1cf938e527/editor/editor_settings.cpp#L352-L355), why not use it?

DeeJayLSP commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -282,27 +314,35 @@ EditorThemeManager::ThemeConfiguration EditorThemeManager::_create_theme_config(
-				preset_contrast = config.default_contrast;
 			} else if (config.preset == "Godot 2") {
 				preset_accent_color = Color(0.53, 0.67, 0.89);
 				preset_base_color = Color(0.24, 0.23, 0.27);
```
Considering this is a legacy classic theme, I believe a `preset_icon_saturation` set to 1 should be here too.

Calinou commented:
> This could use a few more screenshots to get a general idea after the thorough feedback gathered from the comments above.

Some screenshots comparing 4.5.stable with this PR (all at 200% editor scale):

### Dark theme

#### Editor

Before | After
-|-
![Screenshot_20251029_191134](https://github.com/user-attachments/assets/2e6b4f30-400f-487d-ad47-a2309d1e6c17) | ![Screenshot_20251029_190843](https://github.com/user-attachments/assets/ed717f3e-cbbb-4d14-9365-336ec314bb71)

#### Input map dialog

Before | After
-|-
![Screenshot_20251029_191157](https://github.com/user-attachments/assets/511aa574-61b5-4b21-9aeb-17cbd1dfefe2) | ![Screenshot_20251029_190950](https://github.com/user-attachments/assets/eb61d701-4082-49a7-b2e1-792bb1ef461a)

### Light theme

#### Editor

Before | After
-|-
![Screenshot_20251029_191146](https://github.com/user-attachments/assets/9e843db2-e7d4-4d13-8c7f-dc094759c92d) | ![Screenshot_20251029_190902](https://github.com/user-attachments/assets/5a70d7db-ab6f-4fe9-834c-f417b5f63ece)

#### Input map dialog

Before | After
-|-
![Screenshot_20251029_191150](https://github.com/user-attachments/assets/81f8b116-39cd-4bc3-b023-ede5991d0a60) | ![Screenshot_20251029_190940](https://github.com/user-attachments/assets/b3e3b7a7-8397-47e6-8610-d0d85c43ff43)

As you can see, the hover documentation (EditorHelpBit) has trouble distinguishing itself from the rest of the editor, as it doesn't automatically get a drop shadow (not on my Fedora 42 KDE X11 setup at least, even with compositing enabled). We should ensure it always has border somehow, like it did previously.

Mickeon commented:
Still not a fan of the Scene and Filesystem not having any sort of darker background, but maybe it's something worth getting used to.
Actually, I noticed that in the Light theme the lack of a background makes the contents look bleached, unfinished. It's a bit weird.

Giganzo commented:
Input map dialog is a bit harder to see what belongs to what.

akien-mga reviewed (approved):
Great work @YeldhamDev and @passivestar, and everyone reviewing so far!

This seems in a pretty good state to merge now and get more testing. The PR is already quite big and handling all remaining feedback in a big PR is cumbersome.

So we're aware that some feedback is still unaddressed, but let's keep track of it in dedicated issues and fix it in follow-up PRs, so we can fine tune the details and ensure most users are satisfied with the new theme by the time 4.6-stable releases.

Merging this now means that the first version will be available in the upcoming 4.6-dev3 snapshot, but again further work will be done during the dev (and maybe beta) phase to polish it up.

Repiteo commented:
Thanks! Excellent job to everyone involved for bringing this together!

braids commented:
Making a theme default is tacit endorsement of the design of that theme. I don't wish for future design choices to be based off of this theme.

HexDOTexe commented:
Without getting caught up in my own personal nitpicks with this chosen theme (of which I have many), I would hope that function and usability remains the primary factor by which any default theme is selected.

Functionally, there appears to be a loss in effective screen real estate for many elements, as they now take up more space than before. And as pointed out a moment ago, the new Input Map is simply a downgrade. I can't see a good argument for this as it currently exists.

As a positive outlier, I like the new scene and file browser list density, as they both display more items than before in the same space. (Would still prefer a return of a darker background element to help distinguish the item list, though.)

Subjectively, I find this new theme quite difficult to look at. I find this dark brown to be very disagreeable, and I can't recall a time in any software I've ever used where a dark mode theme was a shade of muddy brown. Again, I suppose this is a subjective complaint, but I also would argue that a theme based around a highlighter yellow base would not be a great default choice either.  Personally, I think a color selection closer to the existing default theme would help a lot.

braids commented:
Some of this feedback was already in the thread for a month, like the concerns about the solid-color backgrounds. Why do I have to make a new issue when it was ignored? Additionally, where was it decided that this would be the new theme? [That comment](https://github.com/godotengine/godot/pull/111118#issuecomment-3357623048) still hasn't been addressed.

I don't think it's wise to barrel forward with a large UI change that most of the userbase won't be aware of until it's out! I only just became aware an hour after it was merged.

braids commented:
> @braids we'll make sure to personally ask you before making any change in the engine, my apologies 🫡

Thank you, I left you an issue over here: https://github.com/godotengine/godot/issues/112229

NicholasBolen commented:
Could we have a [project tracker](https://github.com/orgs/godotengine/projects/58) for issues/PRs related to this? I'm not sure how they're usually handled but would find it helpful for seeing progress and preventing overlap.

So far, I've seen:

- https://github.com/godotengine/godot/issues/112234
- https://github.com/godotengine/godot/issues/112229 (incomplete)
- https://github.com/godotengine/godot/issues/112232
   - https://github.com/godotengine/godot/pull/112243

radosmn commented:
Although it looks slick, but the blending of menus and tabs and lists is confusing, I hope in the next version we can differentiate, like in scene tab we can still have trees lines between parent child nodes, and different color gradient to separate each menu or submenus.

rakun99 commented:
<img width="1440" height="860" alt="my_edit_black" src="https://github.com/user-attachments/assets/23305249-95a0-4eea-80d7-6cfb64417e6f" />
Slшghtly edited panels to make more depth.
